### PR TITLE
Add can

### PR DESCRIPTION
**Pull request by Palomino34**

### CODE REVIEW

File: Devices/G80/Device.h [hunk 1]
```diff
@@ -102,4 +102,8 @@
 #define STM32F4_USB_DM_PINS { { PIN(A, 11), AF(10) } }
 #define STM32F4_USB_DP_PINS { { PIN(A, 12), AF(10) } }
 #define STM32F4_USB_VB_PINS { { PIN(A,  9), AF(10) } }
-#define STM32F4_USB_ID_PINS { { PIN(A, 10), AF(10) } }
\ No newline at end of file
+#define STM32F4_USB_ID_PINS { { PIN(A, 10), AF(10) } }
+
+#define INCLUDE_CAN
```
Comment (Arke64) [inline Devices/G80/Device.h]: Move this between ADC and DAC.

File: Targets/STM32F4/STM32F4_Interrupt.cpp [hunk 2]
```diff
@@ -88,6 +88,9 @@ bool STM32F4_InterruptInternal_Activate(uint32_t index, uint32_t *isr, void* isr
     uint32_t *irq_vectors = (uint32_t*)&__Vectors;
 
     irq_vectors[id + 16] = (uint32_t)isr; // exception = irq + 16
+
+    __DMB();
```
Comment (Arke64) [inline Targets/STM32F4/STM32F4_Interrupt.cpp]: Make a separate PR for this and explain why we need it

File: Targets/STM32F4/STM32F4.h [hunk 3]
```diff
@@ -333,4 +333,25 @@ TinyCLR_Result STM32F4_Display_DrawBuffer(const TinyCLR_Display_Provider* self,
 TinyCLR_Result STM32F4_Display_WriteString(const TinyCLR_Display_Provider* self, const char* buffer);
 int32_t STM32F4_Display_GetWidth(const TinyCLR_Display_Provider* self);
 int32_t STM32F4_Display_GetHeight(const TinyCLR_Display_Provider* self);
-TinyCLR_Display_InterfaceType STM32F4_Display_GetType(const TinyCLR_Display_Provider* self);
\ No newline at end of file
+TinyCLR_Display_InterfaceType STM32F4_Display_GetType(const TinyCLR_Display_Provider* self);
+
+const TinyCLR_Api_Info* STM32F4_Can_GetApi();
```
Comment (Arke64) [inline Targets/STM32F4/STM32F4.h]: Match the order in STM32F4.h, this should be between ADC and DAC (alphabetical essentially)

File: Targets/STM32F4/STM32F4_CAN.cpp [hunk 4]
```diff
@@ -0,0 +1,1597 @@
+// Copyright Microsoft Corporation
```
Comment (Arke64) [inline Targets/STM32F4/STM32F4_CAN.cpp]: If this file was not originally copyright Microsoft or Oberon, don't include them here.

File: Targets/STM32F4/STM32F4_CAN.cpp [hunk 5]
```diff
@@ -0,0 +1,1597 @@
...
+static const int TOTAL_CAN_CONTROLLERS = SIZEOF_ARRAY(g_STM32F4_Can_Tx_Pins);
+
+static STM32F4_Can_Controller canController[TOTAL_CAN_CONTROLLERS];
+
+static TinyCLR_Can_Provider *canProvider[TOTAL_CAN_CONTROLLERS];
+static TinyCLR_Api_Info canApi;
+
+static uint8_t canProviderDefs[TOTAL_CAN_CONTROLLERS * sizeof(TinyCLR_Can_Provider)];
+
+void InsertionSort(uint32_t *arr, int32_t length) {
```
Comment (Arke64) [inline Targets/STM32F4/STM32F4_CAN.cpp]: See if InsertionSort and the other three sort methods below exist in the C++ STL. If they don't and we can't change our code to conform to what does, move these functions to a new top level "Shared" folder. Then make a separate PR to add that folder to the compile script.

File: Main/main.cpp [hunk 6]
```diff
@@ -17,6 +17,8 @@
 
 #define TARGET(a) CONCAT(DEVICE_TARGET, a)
 
+const TinyCLR_Api_Provider* globalApiProvider;
```
Comment (Arke64) [inline Main/main.cpp]: Make a separate PR for adding globalApiProvider to main and change its name to "apiProvider".

File: Targets/STM32F4/STM32F4_CAN.cpp [hunk 7]
```diff
@@ -0,0 +1,1597 @@
...
+
+#include <algorithm>
+#include <string.h>
+#include "STM32F4.h"
+
+///////////////////////////////////////////////////////////////////////////////
+
+#ifdef INCLUDE_CAN
+
+#define CAN_MESSAGES_MAX 128
```
Comment (Arke64) [inline Targets/STM32F4/STM32F4_CAN.cpp]: Remove this and replace it with a define in the device.h file, "STM32F4_CAN_RX_BUFFER_DEFAULT_SIZE".

File: Targets/STM32F4/STM32F4_CAN.cpp [hunk 8]
```diff
@@ -0,0 +1,1597 @@
...
+    return TinyCLR_Result::Success;;
+}
+
+TinyCLR_Result STM32F4_Can_GetSourceClock(const TinyCLR_Can_Provider* self, uint32_t &sourceClock) {
+    sourceClock = STM32F4_APB1_CLOCK_HZ;
+
+    return TinyCLR_Result::Success;;
+}
+
+TinyCLR_Result STM32F4_Can_SetReceiveBufferSize(const TinyCLR_Can_Provider* self, int32_t size) {
```
Comment (Arke64) [inline Targets/STM32F4/STM32F4_CAN.cpp]: If there is an internal transmit buffer, it should have a function to control its buffer as well.

File: Devices/G80/Device.h (hunk 1 above)
Comment (Palomino34) [reply Devices/G80/Device.h]: Fixed!
